include/private/gcconfig.h

```diff
@@ -895,7 +895,19 @@ extern char etext[];
 #ifdef HAIKU
 #  define OS_TYPE "HAIKU"
 #  define DYNAMIC_LOADING
-#  define MPROTECT_VDB
+#  define USE_GET_STACKBASE_FOR_MAIN
+#  undef MPROTECT_VDB		/* slow */
+#  ifndef HAVE_PTHREAD_GETATTR_NP
+#    define HAVE_PTHREAD_GETATTR_NP 1
+#  endif
+#  ifndef HAVE_CLOCK_GETTIME
+#    define HAVE_CLOCK_GETTIME 1
+#  endif
+#  ifndef USE_MMAP
+#    define USE_MMAP 1
+#  endif
+#  define USE_MMAP_ANON
+#  define USE_SPIN_LOCK
```

Comment:
> Interestingly with USE_SPIN_LOCK it fails even faster; instead of taking a handful of seconds (with low CPU usage) before crashing, it crashes very quickly after starting with the "Unexpected segmentation fault outside heap ...".

Let me not define USE_SPIN_LOCK. A comment near define USE_SPIN_LOCK in gc_lock.h says:
```
/* In the THREAD_LOCAL_ALLOC case, the allocator lock tends to    */
/* be held for long periods, if it is held at all.  Thus spinning */
/* and sleeping for fixed periods are likely to result in         */
/* significant wasted time.  We thus rely mostly on queued locks. */
```
I think a more detailed investigation is needed whether to use spin locks (and not only for Haiku).

Reply:
It doesn't fail at all without fork(), it works fine with the spin lock, and it's much faster than without (I think 1-2 seconds vs. over 10 for gctest, with vs. without spinlocks.) The spinlock code falls back to a queued lock after a number of spins fail, it appears?

On Haiku, pthread locks never spin but always call into the kernel if the lock can't be acquired with a simple atomic operation. On other OSes, some spins may be used ("PTHREAD_MUTEX_ADAPTIVE" or the like.) `USE_SPIN_LOCK` seems to replicate that behavior.

Reply:
> it's much faster than without (I think 1-2 seconds vs. over 10 for gctest, with vs. without spinlocks.)

I see. Would be nice if you could provide a table of gctest run time (e.g. the one printed by "testing done in %lu ms") of 10-20 runs with and w/o USE_SPIN_LOCK. (I your PC is too powerful, you could increase NTHREADS to make each run longer to reduce noise.)